### PR TITLE
Test external accounts (issue 165)

### PR DESCRIPTION
**Pull request by dabiid**
Tests for issue #165. I'm not sure how general/specific they should be. I created a basic one that doesn't change anything that simply verifies the external accounts and irc nickname forms exist. It's not really needed with the other two but I thought it'd be good to have a smoke test for prd (?). 

I can remove or modify the tests as desired. For example, test_that_user_can_add_external_account could actually add an account rather than just click the link to add a new row and a second test could delete it. Selecting the dropdown with a valid url/name is a bit more complicated than what I have now so I didn't want to get started on it if it's not wanted.

### CODE REVIEW

**Comment by rbillings** [comment]
I took a look at this today- I did see it earlier but was confused by the closure & re-opening. 
First comment: you'll need to update this to use native asserts. You can see how that's done in https://github.com/mozilla/mozillians-tests/commit/f3083b03ec46665a896d8fa670ef47dd20184253

Secondly: I think adding IRC nick is a good start for these, I'll let someone else more familiar state if that's sufficient for the intended smoke tests.

I'll be happy to check this out further after the native asserts update.

**Comment by dabiid** [comment]
I updated this to use native asserts. When I pulled the recent changes from master and rebased, I apparently didn't add a second commit. Sorry about that. I was more careful on the other PR (I at least added a second commit comment, even if most of the changes got wrapped up in the rebasing... I'm still mediocre with git). Point is, this should be ready to look at.

p.s. the reason I closed/opened this is that the CI build got stuck and never returned (it just stopped after like 10 minutes) so I closed/opened rather than push a new commit and it worked.

File: pages/settings.py [hunk 1]
```diff
@@ -200,6 +209,45 @@ def click_find_group_link(self):
...
+                return len(self._root_element.find_elements(*self._account_row_locator))
+
+            def click_add_account(self):
+                self._root_element.find_element(*self._add_account_locator).click()
+
+        class Irc(PageRegion):
+            _irc_nickname_locator = (By.ID, 'id_ircname')
+            _update_locator = (By.ID, 'form-submit-top')
+
+            def get_nickname(self):
```
Comment (davehunt) [inline pages/settings.py]: Make this a `@property` and remove the `get_` prefix.

File: tests/test_profile.py [hunk 2]
```diff
@@ -235,3 +235,50 @@ def test_that_links_in_the_services_page_return_200_code(self, base_url, seleniu
...
+    @pytest.mark.nondestructive
+    def test_that_user_can_view_external_accounts(self, base_url, selenium, vouched_user):
+        home_page = Home(base_url, selenium)
+        home_page.login(vouched_user['email'], vouched_user['password'])
+        settings = home_page.header.click_settings_menu_item()
+
+        irc_form = settings.external_accounts.irc_form
+        external_accounts_form = settings.external_accounts.external_accounts_form
+        assert irc_form
+        assert external_accounts_form
```
Comment (davehunt) [inline tests/test_profile.py]: These asserts are just checking that the page objects are returning something thruthy. I suspect you actually want to verify that these elements are displayed. Adding an `is_displayed` to each page region object might be a good approach. Also, there's not really any need to create variables for these, we can just use them directly.

File: tests/test_profile.py [hunk 3]
```diff
@@ -235,3 +235,50 @@ def test_that_links_in_the_services_page_return_200_code(self, base_url, seleniu
...
+    def test_that_user_can_add_external_account(self, base_url, selenium, vouched_user):
+        home_page = Home(base_url, selenium)
+        home_page.login(vouched_user['email'], vouched_user['password'])
+        settings = home_page.header.click_settings_menu_item()
+
+        external_accounts_form = settings.external_accounts.external_accounts_form
+        cnt_external_accounts = external_accounts_form.count_external_accounts()
+        external_accounts_form.click_add_account()
+        new_cnt_external_accounts = external_accounts_form.count_external_accounts()
+        assert new_cnt_external_accounts == (cnt_external_accounts + 1)
```
Comment (davehunt) [inline tests/test_profile.py]: Nit: We tend to use `assert <expected> <actual>`, so you'll need to switch these around.

File: tests/test_profile.py [hunk 4]
```diff
@@ -235,3 +235,50 @@ def test_that_links_in_the_services_page_return_200_code(self, base_url, seleniu
...
+        settings = home_page.header.click_settings_menu_item()
+
+        irc_form = settings.external_accounts.irc_form
+        old_nickname = irc_form.get_nickname()
+        new_nickname = old_nickname + 'test'
+        irc_form.update_nickname(new_nickname)
+        irc_form.click_update()
+
+        profile_page = home_page.header.click_view_profile_menu_item()
+        assert profile_page.irc_nickname == new_nickname
```
Comment (davehunt) [inline tests/test_profile.py]: As above re expected/actual positions.

File: tests/test_profile.py (hunk 2 above)
Comment (dabiid) [inline tests/test_profile.py]: Good point. I was assuming that if it found the page region without an exception that it basically meant it was displayed. I've made that more explicit. I added it to the ExternalAccountsTab class rather than the two subsections in the page because using those selectors seemed best. Let me know if you think it should be implemented differently.

File: pages/settings.py (hunk 1 above)
Comment (dabiid) [inline pages/settings.py]: Good idea. Done.

File: tests/test_profile.py (hunk 3 above)
Comment (dabiid) [inline tests/test_profile.py]: Oops, I knew that... Updated.

File: tests/test_profile.py [hunk 5]
```diff
@@ -235,3 +235,48 @@ def test_that_links_in_the_services_page_return_200_code(self, base_url, seleniu
...
+
+    @pytest.mark.credentials
+    @pytest.mark.nondestructive
+    def test_that_user_can_view_external_accounts(self, base_url, selenium, vouched_user):
+        home_page = Home(base_url, selenium)
+        home_page.login(vouched_user['email'], vouched_user['password'])
+        settings = home_page.header.click_settings_menu_item()
+
+        assert settings.external_accounts.is_irc_form_present
+        assert settings.external_accounts.is_external_accounts_form_present
```
Comment (davehunt) [inline tests/test_profile.py]: Presence can be misleading as elements can be present but not displayed. You could add an `is_displayed` property to both `ExternalAccounts` and `Irc` which just returns `self._root_element.is_displayed()`.

**Comment by davehunt** [comment]
r+ wc

**Comment by dabiid** [comment]
@davehunt is this one ready to be squashed as well? I'm not sure if you saw the recent changes. And what does r+ wc mean (I found r+ described [here](https://wiki.mozilla.org/Firefox/Code_Review))

**Comment by stephendonner** [comment]
(r+ wc == review+ with comments [addressed])

**Comment by davehunt** [comment]
Needs rebasing and squasing, then it can be merged. Please comment when you push changes and are ready for review, as pushing alone doesn't trigger a notification.

**Comment by dabiid** [comment]
@stephendonner thanks for the explanation. @davehunt I didn't realize the push alone didn't send a notification, that's good to know. Commits squashed.

**Comment by davehunt** [comment]
@dabiid - I have a failure when running one of the new tests:

```
self = <tests.test_profile.TestProfile instance at 0x10765bfc8>, base_url = 'https://mozillians.allizom.org'
selenium = <selenium.webdriver.firefox.webdriver.WebDriver (session="1fd6fa58-8786-cb40-9c19-fed9a2bad069")>
vouched_user = {'email': 'webqabrowserid@aol.com', 'name': 'Mozillians.User', 'password': 'Password123'}

    @pytest.mark.credentials
    def test_that_user_can_modify_external_accounts_irc_nickname(self, base_url, selenium, vouched_user):
        home_page = Home(base_url, selenium)
        home_page.login(vouched_user['email'], vouched_user['password'])
        settings = home_page.header.click_settings_menu_item()
    
        irc_form = settings.external_accounts.irc_form
        old_nickname = irc_form.nickname
        new_nickname = old_nickname + 'test'
        irc_form.update_nickname(new_nickname)
        irc_form.click_update()
    
        profile_page = home_page.header.click_view_profile_menu_item()
>       assert new_nickname == profile_page.irc_nickname
E       assert 'tetetetest' == 'tetetete'
E         - tetetetest
E         ?         --
E         + tetetete
```

Each time I run it the expected value increases - the previous run was 'tetetest'.

**Comment by dabiid** [comment]
@davehunt Well that sucks. I was encountering something like that before I added the element.clear() call. It looks like some combination of the clear/send timing which I don't understand... It's passing for me with my user and with the vouched user credentials that you had so that's not helping me figure it out. Do you think adding an explicit wait of some kind would work?

Also, that test will fail if the nickname starts out empty. Is it worth handling that? It'd take something like this:

    @property
    def irc_nickname(self):
        try:
            return self.selenium.find_element(*self._irc_nickname_locator).text
        except NoSuchElementException:
            return ''

This shouldn't mask the selector breaking because the first assert would fail in that case (new_nickname == '').

**Comment by davehunt** [comment]
Ah, I remember this one now! Mozillians has a keyboard shortcut 's' that moves focus to the search box, to typing a string that contains 's' causes problems. If you change your string to be a timestamp or similar it should work. Of course, if the test fails it may mean the nick is not changed back, and could become a very long string due to the concatenation.

Ideally we'd be using a new user for each test. We can do this with the FxA test accounts, however we'll need to somehow vouch and register these users programmatically. For now, I'd be happy with appending a single character, but longer term we should address this issue properly.

**Comment by dabiid** [comment]
@davehunt Interesting, I had no idea about the s. I choose "_" for no reason in particular (other than that it's a bit unusual). 

I didn't add the changes to handle a null nickname so let me know if I should add that.

And finally, let me know if this looks good and I'll squash the commits.

**Comment by davehunt** [comment]
Looks good. I'll rebase, squash, and merge.

**Comment by davehunt** [comment]
Ah, I can't do that because we need a Travis CI run to pass before merging. Could you take care of the rebase and squash, @dabiid?

**Comment by dabiid** [comment]
@davehunt I updated with the changes from master and rebased so it should be good to go. 

Also, should I submit the PR for issue187 again? Maybe with a comment saying it needs an unvouched user in the envs but has been reviewed in PR #209.

**Comment by davehunt** [comment]
@dabiid looks like it needs rebasing from upstream master again. As for #187 please do open a new pull, but could you also provide an email/password for an unvouched user that we can use in our CI.

**Comment by dabiid** [comment]
@davehunt - I'll take care of the rebasing shortly. As for the user, would it be okay if I create something like webqabrowserid_unvouched@zoho.com (does the host matter)? I've been using personal emails for testing.

**Comment by davehunt** [comment]
@dabiid the host does not matter, so you should be fine with restmail, mailinator, zoho, etc.

**Comment by dabiid** [comment]
@davehunt ok, this should be ready. I'll update branch 187 and submit a new PR with a user. Thanks again for your help on this.
